### PR TITLE
#163187456 Add user authentication with jwt 

### PR DESCRIPTION
**Pull request by tosinibrahim96**
#### Description of Task to be completed?
When a user registers on the system
Then the user should receive a JWT after successful registration.

When a user logs into the system
Then the user should receive a JWT after successful login.

#### How should this be manually tested?
1. run `npm install`
2. on postman, use the url `localhost:3000/api/v1/auth/signup` for adding a user and `localhost:3000/api/v1/auth/signin` to log in a user

#### Any background context you want to provide?
A  database must be available for `sequelize` to store data. See the env-sample files for database information
`sequelize-cli` was added for travis build
dialect must be explicitly specified for travis build

#### What are the relevant pivotal tracker stories?
[As a system, I should be able to register and log in users]
(https://www.pivotaltracker.com/story/show/163187456)

#### Screenshots 
This happens when a secret key is not set on travis site
screen shot 2019-01-17 at 8 23 49 am

You need to signup and go to settings, then set the secret  key for token generation
screen shot 2019-01-17 at 8 26 20 am

And voila!!!
screen shot 2019-01-17 at 8 20 41 am

### CODE REVIEW

File: controllers/AuthController.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+  /**
+   * Create a User.
+   * @param {object} req - The request sent by user.
+   * @param {object} res - The request sent by user.
+   * @param {function} next - The next function to send email to user.
+   * @returns {object} The status of response and body of the response.
+   */
+  static async signUp(req, res, next) {
+    try {
+      // Hash password before sending
```
Comment (codrex) [inline controllers/AuthController.js]: Please remove inline comments

File: controllers/AuthController.js [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+      const hashedPassword = await bcrypt.hash(req.body.password, 8);
+      const queryResult = await userModelManager.create(
```
Comment (codrex) [inline controllers/AuthController.js]: consider changing `queryResult` to `userRecord` because the operation being performed by `userModelManager.create` is not a query operation

File: controllers/AuthController.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        hashedPassword,
+        req.body.firstname,
+        req.body.lastname
+      );
+
+      // Get Id of the new user and generate a token
+      const { id } = queryResult.dataValues;
+
+      // Attach token and mail to request body
+      req.jwtToken = jwtUtil.generateToken('1d', id);
```
Comment (codrex) [inline controllers/AuthController.js]: typo here `'1d'`

File: controllers/AuthController.js [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+  /**
+   * Login a registered user.
+   * @param {object} req - The request sent by user.
+   * @param {object} res - The response sent to user.
+   * @returns {object} The status of response and body of the response.
+   */
+  static async signIn(req, res) {
+    try {
+      const { id } = req.userDetails;
```
Comment (codrex) [inline controllers/AuthController.js]: Please put me through what is going on here
`const { id } = req.userDetails;`

File: controllers/AuthController.js [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+      const token = jwtUtil.generateToken('1d', id);
```
Comment (codrex) [inline controllers/AuthController.js]: typo

File: controllers/AuthController.js [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+      return res.status(200).json({
+        success: true,
+        message: 'Login Successful',
+        login_token: token
```
Comment (codrex) [inline controllers/AuthController.js]: why do we have snake case here? `  login_token`

File: database/config.js [hunk 7]
```diff
@@ -13,7 +13,7 @@ module.exports = {
     password: process.env.DATABASE_PASSWORD,
     database: process.env.TEST_DATABASE_NAME,
     host: process.env.DATABASE_HOST,
-    dialect: process.env.DATABASE_DIALECT
+    dialect: 'postgres'
```
Comment (codrex) [inline database/config.js]: why the change here?

File: lib/modelManagers/usermodel.js [hunk 8]
```diff
@@ -44,6 +47,22 @@ class User {
...
+   *
+   * @param {field} field The field to be searched.
+   * @static
+   * @memberof User
+   * @returns {object} The parameters from the query
+   */
+  static async findUser(field) {
+    const UserRecord = await user.findOne({
+      where: {
+        email: field
```
Comment (codrex) [inline lib/modelManagers/usermodel.js]: Why are we finding a user just by email?

File: middlewares/jwtValidator.js [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+  // Get the token sent from anywhere(i.e header,body or query)
+  const token = req.body.token || req.query.token || req.headers.token || req.params;
+  if (token) {
+    // Decode token if it is not empty and call next function
+    try {
+      req.decodedToken = jwtUtil.verifyToken(token);
+      next();
+    } catch (err) {
+      // Return an error if it occurs
+      res.json({
```
Comment (codrex) [inline middlewares/jwtValidator.js]: What is the status code being sent to the user here?

File: middlewares/jwtValidator.js [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+        success: false,
+        message: err.message
```
Comment (codrex) [inline middlewares/jwtValidator.js]: Explicitly stating the error that occurred here is a better way to go as it gives us the ability to clearly communicate with the user.

File: routes/api/index.js [hunk 11]
```diff
@@ -13,4 +13,6 @@ router.use((err, req, res, next) => {
   return next(err);
 });
 
+router.use('/auth/', require('../api/authRoutes/authRoutes'));
```
Comment (codrex) [inline routes/api/index.js]: `'../api/authRoutes/authRoutes'` is not cool enough, please make it `'../api/authRoutes'`

File: test/auth/auth.test.js [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+        lastname: 'testlastname'
+      }
+    };
+    const correctDetails = {
+      email: 'testmail@gmail.com',
+      password: '111111'
+    };
+
+    it('Should add a new user to the database', async () => {
+      await authController.signUp(req, res, next);
```
Comment (codrex) [inline test/auth/auth.test.js]: what are we asserting for here?
` await authController.signUp(req, res, next);`

File: test/auth/auth.test.js [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+    });
+    it('Should check if added user exist', async () => {
+      newUser = await userModelManager.findUser(req.body.email);
```
Comment (codrex) [inline test/auth/auth.test.js]: `newUser` local variable is more prefered over a global variable as it makes code easy to follow and maintainable.

File: test/modelmanagers/usermodel.test.js [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+let createdUser;
+
+describe('Test for sending user info into database', () => {
+  it('Should have method that sends user info to database', () => {
+    expect(userModelManager.create).to.be.a('function');
+  });
+
+  it('should return a creted user object', async () => {
+    before(() => {
+      expect(firstname).to.be.a('string');
```
Comment (codrex) [inline test/modelmanagers/usermodel.test.js]: The `before` function is not met for assertions.

File: test/utils/jwtUtil.test.js [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+let token, verifiedToken;
+
+describe('Test for creation and verification of token', () => {
+  it('Should have methods to generate and verify token', () => {
+    expect(jwtUtil.generateToken).to.be.a('function');
+    expect(jwtUtil.verifyToken).to.be.a('function');
+  });
+
+  it('should return generated token and decode the token', () => {
+    before(() => {
```
Comment (codrex) [inline test/utils/jwtUtil.test.js]: I would love to know why you are making assertions in a `before` function.

**Review by codrex — COMMENTED** [review]
Nice work so far.

Please move all the string literals and constants in your work to the constants file.

Please attend to my comments.

File: controllers/AuthController.js (hunk 1 above)
Comment (tosinibrahim96) [reply controllers/AuthController.js]: I am checking them out and removing now

File: controllers/AuthController.js (hunk 3 above)
Comment (tosinibrahim96) [reply controllers/AuthController.js]: The string here is one day. The first parameter is valid time of the jwt.

File: controllers/AuthController.js (hunk 2 above)
Comment (tosinibrahim96) [reply controllers/AuthController.js]: Done

File: controllers/AuthController.js (hunk 4 above)
Comment (tosinibrahim96) [reply controllers/AuthController.js]: This is the explanation i was giving in the morning. The function has changed now based on your clarifications. It is not valid

File: controllers/AuthController.js (hunk 5 above)
Comment (tosinibrahim96) [reply controllers/AuthController.js]: As i said in the first comment, String one day. It is not a typo

File: controllers/AuthController.js (hunk 6 above)
Comment (tosinibrahim96) [reply controllers/AuthController.js]: I will change that immediately

File: database/config.js (hunk 7 above)
Comment (tosinibrahim96) [reply database/config.js]: Travis was complaining that it has to be hard-coded. I have included it in the .yml file and changed this back to process.env.DATABASE_DIALECT

File: lib/modelManagers/usermodel.js (hunk 8 above)
Comment (tosinibrahim96) [reply lib/modelManagers/usermodel.js]: The function checks if the email is available during login.

File: middlewares/jwtValidator.js (hunk 9 above)
Comment (tosinibrahim96) [reply middlewares/jwtValidator.js]: I have included the response code

File: middlewares/jwtValidator.js (hunk 10 above)
Comment (tosinibrahim96) [reply middlewares/jwtValidator.js]: The err is coming from jtw library and it has a message property that describes what error that has occured. That's why i did that

File: routes/api/index.js (hunk 11 above)
Comment (tosinibrahim96) [reply routes/api/index.js]: i'll change that

File: test/auth/auth.test.js (hunk 12 above)
Comment (tosinibrahim96) [reply test/auth/auth.test.js]: I'm testing if a new user is added to the database that's why i'm immediately checking if the added user exist after running that function

File: test/auth/auth.test.js (hunk 13 above)
Comment (tosinibrahim96) [reply test/auth/auth.test.js]: i have changed that

File: test/modelmanagers/usermodel.test.js (hunk 14 above)
Comment (tosinibrahim96) [reply test/modelmanagers/usermodel.test.js]: Thanks for that info. I have made the changes

File: test/utils/jwtUtil.test.js (hunk 15 above)
Comment (tosinibrahim96) [reply test/utils/jwtUtil.test.js]: I just read about it. The info on mocha docs said it runs that block before anything. I did not realise it's not for assertions. correction taken

File: test/auth/auth.test.js (hunk 12 above)
Comment (eltNEG) [reply test/auth/auth.test.js]: @tosinibrahim96 you can move `await authController.signUp(req, res, next); to a 'before' function instead performing the operation in 'it'.

File: middlewares/jwtValidator.js (hunk 10 above)
Comment (eltNEG) [reply middlewares/jwtValidator.js]: @tosinibrahim96 Add a `return` before `next()` and also before the `res.status` after catch.

File: test/auth/auth.test.js (hunk 12 above)
Comment (tosinibrahim96) [reply test/auth/auth.test.js]: i have done that @eltNEG

File: middlewares/jwtValidator.js (hunk 10 above)
Comment (tosinibrahim96) [reply middlewares/jwtValidator.js]: done @eltNEG thanks

File: controllers/AuthController.js (hunk 3 above)
Comment (codrex) [reply controllers/AuthController.js]: cool

File: test/auth/auth.test.js [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+    it('Should send user not found message', async () => {
+      const res = await chai
+        .request(server)
+        .post('/api/v1/users/signin/')
+        .send(incorrectMail);
+      expect(res).to.have.status(404);
+      res.body.success.should.equal(false);
+      res.body.message.should.equal('Email does not match our record');
+    });
+    // it('Should login a verified user', async () => {
```
Comment (codrex) [inline test/auth/auth.test.js]: remove dead code

File: test/utils/jwtUtil.test.js [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+  userId = 'd4d6884c-ed49-4158-99cd-6305af05a4a6';
+let token, verifiedToken;
+
+describe('Test for creation and verification of token', () => {
+  it('Should have methods to generate and verify token', () => {
+    expect(jwtUtil.generateToken).to.be.a('function');
+    expect(jwtUtil.verifyToken).to.be.a('function');
+  });
+
+  it('should return generated token and decode the token', () => {
```
Comment (codrex) [inline test/utils/jwtUtil.test.js]: You are testing for more than one functionality here. Please consider splitting the test

File: lib/modelManagers/usermodel.js (hunk 8 above)
Comment (codrex) [reply lib/modelManagers/usermodel.js]: My expectation here is that `findUser` returns a user's record base on a given query. Please refactor the function to accommodate that functionality

File: routes/api/authRoutes.js [hunk 18]
```diff
@@ -0,0 +1,7 @@
+const router = require('express').Router();
```
Comment (codrex) [inline routes/api/authRoutes.js:1]: Add documentation to your routes

Comment (tosinibrahim96) [reply routes/api/authRoutes.js:1]: No problem. I'll do that

File: test/auth/auth.test.js (hunk 16 above)
Comment (tosinibrahim96) [reply test/auth/auth.test.js]: Yes, i'll do that

File: test/utils/jwtUtil.test.js (hunk 17 above)
Comment (tosinibrahim96) [reply test/utils/jwtUtil.test.js]: I'll do that

File: lib/utils/jwtUtil.js [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+      process.env.SECRET_KEY,
+      { expiresIn: duration }
+    );
+
+    return token;
+  }
+
+  /**
+   * @param {string} userToken
+   * @returns {Object | false} userId after decoding the token or error message
```
Comment (codrex) [inline lib/utils/jwtUtil.js]: I don't see this function ever returning false

File: middlewares/jwtValidator.js (hunk 10 above)
Comment (codrex) [reply middlewares/jwtValidator.js]: Explicitly stating the error that occurred here is a better way to go as it gives us the ability to clearly communicate with the user.

File: middlewares/jwtValidator.js [hunk 20]
```diff
@@ -0,0 +1,31 @@
...
+      return res.status(BAD_REQUEST_CODE).json({
+        success: UNSUCCESSFUL,
+        message: err.message
+      });
+    }
+  }
+
+  return res.status(BAD_REQUEST_CODE).json({
+    success: UNSUCCESSFUL,
+    message: 'No token provided'
```
Comment (codrex) [inline middlewares/jwtValidator.js]: move to constants

File: controllers/AuthController.js [hunk 21]
```diff
@@ -0,0 +1,87 @@
...
+        req.body.lastname
+      );
+
+      const { id } = userRecord.dataValues;
+
+      req.jwtToken = jwtUtil.generateToken(TOKEN_TIMESPAN, id);
+      req.email = req.body.email;
+      return next();
+    } catch (error) {
+      throw error;
```
Comment (codrex) [inline controllers/AuthController.js]: Where is this error been handled?

File: test/middlewares/jwtValidator.test.js [hunk 22]
```diff
@@ -0,0 +1,79 @@
...
+    };
+
+    const res = {
+      status() {},
+      json() {}
+    };
+    const next = sinon.stub();
+    sinon.stub(res, 'status').returnsThis();
+    jwtValidator(req, res, next);
+    // eslint-disable-next-line no-unused-expressions
```
Comment (codrex) [inline test/middlewares/jwtValidator.test.js]: // eslint-disable-next-line no-unused-expressions ??

File: test/middlewares/jwtValidator.test.js [hunk 23]
```diff
@@ -0,0 +1,79 @@
...
+
+    const res = {
+      status() {},
+      json() {}
+    };
+    const next = sinon.stub();
+    sinon.stub(res, 'status').returnsThis();
+    jwtValidator(req, res, next);
+    expect(res.status).to.have.been.calledWith(400);
+    expect(res).to.be.an('object');
```
Comment (codrex) [inline test/middlewares/jwtValidator.test.js:49]: `expect(res).to.be.an('object');`

Please test for the structure of the object

File: lib/utils/jwtUtil.js (hunk 19 above)
Comment (tosinibrahim96) [reply lib/utils/jwtUtil.js]: sorry, that was a mistake. I'll make the changes

File: middlewares/jwtValidator.js (hunk 10 above)
Comment (tosinibrahim96) [reply middlewares/jwtValidator.js]: ok. I will do that

File: middlewares/jwtValidator.js (hunk 20 above)
Comment (tosinibrahim96) [reply middlewares/jwtValidator.js]: I'll do that

File: test/middlewares/jwtValidator.test.js (hunk 22 above)
Comment (tosinibrahim96) [reply test/middlewares/jwtValidator.test.js]: expect(next).to.have.been.calledOnce;....This sinon syntax line always throws eslint error and prevents commit

File: test/middlewares/jwtValidator.test.js (hunk 23 above)
Comment (tosinibrahim96) [reply test/middlewares/jwtValidator.test.js:49]: no problem

File: test/middlewares/jwtValidator.test.js (hunk 22 above)
Comment (tosinibrahim96) [reply test/middlewares/jwtValidator.test.js]: I have changed that now

**Review by codrex — APPROVED** [review]
